service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/transport/amqps/AmqpConnectionHandler.java

```diff
@@ -0,0 +1,487 @@
+    {
+        Link link = event.getLink();
+        if (event.getLink().getName().equals(tag))
+        {
+            Target t = new Target();
+            t.setAddress(endpoint);
+            Source source = new Source();
+            source.setAddress(endpoint);
+            link.setTarget(t);
+            link.setSource(source);
+        }
+    }
+
+    @Override
+    public void onLinkRemoteOpen(Event event)
+    {
+        log.debug("Link with name {} opened remotely", event.getLink().getName());
+        openLinksLatch.countDown();
+    }
+
+    @Override
+    public void onConnectionRemoteOpen(Event event)
+    {
+        log.debug("Connection opened remotely");
+    }
+
+    @Override
+    public void onSessionRemoteOpen(Event event)
+    {
+        log.debug("Session opened remotely");
+    }
+
+    @Override
+    public void onConnectionRemoteClose(Event event)
+    {
+        super.onConnectionRemoteClose(event);
+        event.getTransport().close_tail();
+        releaseOpenLatch();
+    }
+
+    @Override
+    public void onLinkRemoteClose(Event event)
+    {
+        super.onLinkRemoteClose(event);
+        event.getTransport().close_tail();
+        releaseOpenLatch();
+    }
+
+    @Override
+    public void onSessionRemoteClose(Event event)
+    {
+        super.onSessionRemoteClose(event);
+        event.getTransport().close_tail();
+        releaseOpenLatch();
+    }
+
+    @Override
+    public void onLinkLocalClose(Event event)
+    {
+        log.debug("Closing amqp session");
+
+        //Reactor should respond with onSessionLocalClose call, where we close the connection. Cannot close the connection here
+        // because it can only be done when all of its sessions have closed
+        this.connection.session().close();
+    }
+
+    @Override
+    public void onSessionLocalClose(Event event)
+    {
+        log.debug("Closing amqp connection");
+
+        //Reactor should respond with onConnectionLocalClose call, where we close the reactor. Cannot close the reactor here
+        // because it can only be done when all of its connections have closed
+        this.connection.close();
+    }
+
+    @Override
+    public void onConnectionLocalClose(Event event)
+    {
+        log.trace("Stopping reactor");
+        reactor.stop();
+    }
+
+    @Override
+    public void onTransportError(Event event)
+    {
+        super.onTransportError(event);
+        event.getTransport().close_tail();
+        releaseOpenLatch();
+    }
+
+    /**
+     * If an exception was encountered while opening the AMQP connection, this function shall throw that saved exception
+     * @throws IOException if an exception was encountered while opening the AMQP connection. The encountered
+     * exception will be the inner exception
+     * @throws IotHubException if an Iot Hub level exception occurs during transport
+     */
+    protected void validateConnectionWasSuccessful() throws IOException, IotHubException
```

Comment:
This seems to be called from the `SendHandler` only, shouldn't this be called from other clients as well? before performing the operation? Do we do anything with the `savedException`?

Reply:
This should be called by all amqp connections at the end of the open() call, I'll refactor this. Good catch